### PR TITLE
Add lifestyle profile, distance calculating algorithm

### PR DESCRIPTION
**Pull request by gzyon**

### CODE REVIEW

File: src/api/models/lifestyle.py [hunk 1]
```diff
@@ -5,8 +5,120 @@
...
+    like_social: bool = None
+    like_music: bool = None
+    interests: List[str] = None
+    intended_pillar: str = None
+    diet: str = None
+    smoking: str = None
+    preferred_roommate: str = None
+    blacklist_roommate: List[str] = None
+
+@validator("sleep_time", pre=True, always=True)
```
Comment (MarkHershey) [inline src/api/models/lifestyle.py:30]: The indentation from this line onwards(below) seems wrong.

File: src/api/models/lifestyle.py [hunk 2]
```diff
@@ -5,8 +5,120 @@
...
+    if isinstance(v, str) and v.upper() in (
+        "BEFORE MIDNIGHT",
+        "AROUND MIDNIGHT",
+        "AFTER MIDNIGHT",
+    ):
+        return v
+    else:
+        return "AROUND MIDNIGHT"
+
+@validator("social_habits", pre=True, always=True)
```
Comment (MarkHershey) [inline src/api/models/lifestyle.py:41]: where is this field? `social_habits`?

File: src/api/models/lifestyle.py [hunk 3]
```diff
@@ -5,8 +5,120 @@
...
+    if isinstance(v, str) and v.upper() in (
+        "ISTD",
+        "EPD",
+        "ESD",
+        "DAI",
+        "ASD",
+    ):
+        return v
+    else:
+        return "AROUND MIDNIGHT"
```
Comment (MarkHershey) [inline src/api/models/lifestyle.py:106]: ??

File: src/api/models/lifestyle.py [hunk 4]
```diff
@@ -5,8 +5,120 @@
...
+
+@validator("social_habits", pre=True, always=True)
+def validate_social_habits(cls, v):
+    if isinstance(v, bool) or v is None:
+        return v
+    else:
+        return None 
+
+@validator("room_cleanliness", pre=True, always=True)
```
Comment (MarkHershey) [inline src/api/models/lifestyle.py:48]: Ok, seems like you did not understand how this `@validator` works. 
Here is the documentation: https://pydantic-docs.helpmanual.io/usage/validators/

**Review by MarkHershey — CHANGES_REQUESTED** [review]

File: src/api/models/lifestyle.py (hunk 2 above)
Comment (gzyon) [reply src/api/models/lifestyle.py:41]: like_social bool

File: src/api/models/lifestyle.py [hunk 5]
```diff
@@ -3,10 +3,210 @@
...
+    like_clean: int # how tidy do they need their room to be? range from 1 to 5
+    like_music: bool # do they like to have music in the room?
+    relationship_scale: int # from 1 to 5
+    aircon_usage: str # choice (often, occasionally, not at all) 
+    interests: List[str] = [] # select from given list, let them add on if they want (others)
+    intended_pillar: str = "NA" # choice (ISTD, EPD, ESD, ASD, DAI, NA)
+    diet: str = "None" # choice (halal, no beef, kosher, others)
+    smoking: str # choice (social smoker, not smoker, active smoker)
+
+    @validator("sleep_time", pre=True, always=True)
```
Comment (MarkHershey) [inline src/api/models/lifestyle.py:21]: Check again

**Review by MarkHershey — CHANGES_REQUESTED** [review]

**Comment by MarkHershey** [comment]
Pls write a unit test for testing this data model, similar to this example: https://github.com/MarkHershey/SUTDHousingPortal/blob/master/tests/unit_tests/test_model_room_profile.py

**Review by MarkHershey — APPROVED** [review]

**Pull request merged**